ADD command

Vanethos opened the pull request:
Added Add command with the following synthax:

### add

*Aliases: a*

Adds the package to your `pubspec.yaml` file.

`pubx add {package-name}`

Additional options:

* `dev` - add this as a `dev_dependencies`
* `lock` - lock the version using the `^` symbol

Added to README instruction on using this command

Vanethos committed:
Add add command

shyndman commented on lib/src/add.dart:
```diff
@@ -0,0 +1,79 @@
+
+  Future<File> _getPubspec() async {
+    Directory current = Directory.current;
+    var slash = Platform.isWindows ? "\\" : "/";
```
Better to use the path library's join function. It will handle platform differences for you.

Vanethos replied:
Will do!

shyndman commented on lib/src/add.dart:
```diff
@@ -0,0 +1,79 @@
+    );
+    argParser.addFlag(
+      'lock',
+      help: 'Lock version with "^" symbol',
```
Carets don't lock the version — they specify a lower bound in a version range, which ends at the next breaking version (exclusive)

`^1.2.3` is the same as `>=1.2.3 <2.0.0`

https://dart.dev/tools/pub/dependencies#caret-syntax

shyndman replied:
One idea is that we make ^ the default. If you specify --lock, it adds the exact version string, no more, no less.

`pubx add --lock quiver` writes `quiver: 2.0.3`

`pubx add quiver` writes `quiver: ^2.0.3`

Vanethos replied:
👌 seems great to me

shyndman commented on lib/src/add.dart:
```diff
@@ -0,0 +1,79 @@
+            contents.write(line+"\n");
+      }, // Add line to our StringBuffer object
+          onDone: () => pub.writeAsStringSync(contents.toString()),
+          onError: (e) => print('There was a problem adding the dependency to your project'));
```
Better to let this throw. It won't be pretty output, but it might give the user a clue about what went wrong. Plus it gives people a stack to send to us. :)

Vanethos replied:
Will do the rethrow like in the other one :D

shyndman commented on lib/src/add.dart:
```diff
@@ -0,0 +1,79 @@
+    Directory current = Directory.current;
+    var slash = Platform.isWindows ? "\\" : "/";
+    var file = File("${current.path}${slash}pubspec.yaml");
+    if (await file.exists()) {
```
It's not a huge deal, and you can leave it if you want, but in a CLI you're pretty safe to call the synchronous file system methods. There isn't any other code waiting on the thread in this case.

Also, there's actually a performance hit when using some of the file async methods (which is irrelevant here, because you're doing it once, but you might find it interesting). There's even a lint for it:
https://dart-lang.github.io/linter/lints/avoid_slow_async_io.html

Vanethos replied:
This is a very interesting point! However, the if statement does not allow me to do that since it's expecting a `bool` and that value is a `Future<bool>`

shyndman replied:
Try this: https://api.dartlang.org/stable/2.4.0/dart-io/FileSystemEntity/existsSync.html

Vanethos replied:
Oh, I feel so dumb for not checking this ahahaha

shyndman commented on lib/src/add.dart:
```diff
@@ -0,0 +1,79 @@
+    var pub = await _getPubspec();
+
+    if (pub == null) {
+      print("No pubspec.yaml file found, please run this command in the root folder of a Dart project");
```
This seems like a failure mode, doesn't it? Maybe it's worth throwing an exception, and letting someone higher up the stack print the error message, and exit with an error code.

Vanethos replied:
I did as you said, but since this might be a common exception, does it make sense to have the stack printed?

shyndman replied:
I agree, no stack printed. Just the error message and exiting non-zero code.

shyndman commented on lib/src/add.dart:
```diff
@@ -0,0 +1,79 @@
+          onError: (e) => print('There was a problem adding the dependency to your project'));
+
+    } on PackageNotFoundException catch (e) {
+      stderr.writeln('${e.packageName} not found');
```
Also a failure mode. We'll want a non-zero error code here too.

shyndman replied:
We could let this exception bubble up to main, couldn't we? It's already got the catch, the `print()` and the `exit()`.

Vanethos replied:
Changed! :D

shyndman reviewed (changes requested):
Nice work! A few comments.

Vanethos committed:
Add requested changes to PR

Vanethos commented:
@shyndman do you want me to "Resolve conversation" or keep them open?

Vanethos committed:
Change Add method to exit if no pubspec is found, without showing a stacktrace
Change logic behind --lock so that every package is added with the caret ^ symbol by default
Change file.exists to file.existsSync so that the process is synchronous

shyndman commented:
The way that I've learned with code review is that if the reviewee thinks they're done with a comment, they resolve it. If the reviewer disagrees, they reopen on the next pass. It's generally a pretty smooth process.

Just left a little nit. Resolve and we're good to go.

Also, really looking forward to using this :)

Vanethos committed:
Change location of catch for FileNotFoundException

Vanethos committed:
Change location of AddCommand to prevent merge conflicts

Vanethos commented:
All should be done now! (Should have comented this earlier!)

shyndman commented:
I had to merge it by hand. It's now pushed to master, and I'll publish in a few. Thanks again!
